contracts/src/v0.8/ccip/pools/SiloedLockReleaseTokenPool.sol

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.24;
+
+import {ITypeAndVersion} from "../../shared/interfaces/ITypeAndVersion.sol";
+
+import {Pool} from "../libraries/Pool.sol";
+import {TokenPool} from "./TokenPool.sol";
+
+import {IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
+import {EnumerableSet} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/structs/EnumerableSet.sol";
+
+/// @notice A variation on Lock Release token pools where liquidity is shared among some chains, and stored independently
+/// for others. Chains which do not share liquidity are known as siloed chains.
+/// @dev One token per LockReleaseTokenPool.
+contract SiloedLockReleaseTokenPool is TokenPool, ITypeAndVersion {
+  using SafeERC20 for IERC20;
+  using EnumerableSet for EnumerableSet.UintSet;
+
+  error InsufficientLiquidity();
+  error LiquidityNotAccepted();
+
+  event LiquidityTransferred(uint64 remoteChainSelector, address indexed from, uint256 amount);
+  event LiquidityAdded(uint64 remoteChainSelector, address indexed provider, uint256 indexed amount);
+  event LiquidityRemoved(uint64 remoteChainSelector, address indexed provider, uint256 indexed amount);
+  event ChainSiloeDesignationUpdated(uint64 remoteChainSelector, bool isSiloed);
+
+  event RebalancerSet(uint64 indexed remoteChainSelector, address oldRebalancer, address newRebalancer);
+
+  string public constant override typeAndVersion = "SiloedLockReleaseTokenPool 1.5.10-dev";
+
+  mapping(uint64 chainSelector => uint256 lockedBalance) internal s_lockedTokensByChainSelector;
+  mapping(uint64 remoteChainSelector => address rebalancer) internal s_rebalancerByChain;
+  mapping(uint64 remoteChainSelector => bool shouldSiloFunds) internal s_siloedChainSelectors;
```

Comment:
These are three mappings all with the same key. This is probably not the most gas efficient solution